dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java

```diff
@@ -183,61 +179,111 @@ public void setInterceptorChain(InterceptorChain ic) {
 
     @Override
     public Object put(String key, Object value) {
-        if (contextCache != null) {
-            contextCache.put(key, value);
-        }
         return super.put(key, value);
     }
 
     @Override
     public Object getContextualProperty(String key) {
-        if (contextCache == null) {
-            calcContextCache();
+        Object o = super.get(key);
+        if (o != null) {
```

Comment:
I have some concern that with the cache mechanism there could be a null value that would override set values in the other property tiers.  If that is possible this null check and the null checks in getFromExchange is not sufficient.  You would need to also check to see if containsKey() if null is a valid value.

Reply:
used getOrDefault API to replicate previous behavior